pkg/user/tx_client.go

```diff
@@ -199,19 +208,25 @@ func SetupTxClient(
 
 // SubmitPayForBlob forms a transaction from the provided blobs, signs it, and submits it to the chain.
 // TxOptions may be provided to set the fee and gas limit.
-func (client *TxClient) SubmitPayForBlob(ctx context.Context, blobs []*blob.Blob, opts ...TxOption) (*sdktypes.TxResponse, error) {
+func (client *TxClient) SubmitPayForBlob(ctx context.Context, blobs []*blob.Blob, opts ...TxOption) (*TxResponse, error) {
 	resp, err := client.BroadcastPayForBlob(ctx, blobs, opts...)
-	if err != nil {
-		return resp, err
+	if err != nil && resp != nil {
+		return &TxResponse{Code: resp.Code, TxHash: resp.TxHash}, fmt.Errorf("failed to broadcast pay for blob: %v", err)
+	} else if err != nil {
+		return &TxResponse{}, fmt.Errorf("failed to broadcast pay for blob: %v", err)
 	}
```

Comment:
this kind of confuses me. It seems like an anti-pattern to use the response if `err != nil`. Are these lines necessary or can they be removed in favor of the more common pattern:

```suggestion
	if err != nil {
		return &TxResponse{}, fmt.Errorf("failed to broadcast pay for blob: %v", err)
	}
```


If the extra lines are necessary, can this PR include a unit test for the scenario that covers those lines? Why is the information in the error insufficient?

Reply:
depending on at what point the tx failed the response will include tx hash and error code. Previously we were directly returning the response from BroadcastPayForBlob if it failed. I had to add `resp != nil` check because i no longer have a shared response type for BroadcastPayForBlob and ConfirmTx which means that I'd have to construct the  response myself and it'll panic if response is nil.

I'd say TxClient tests are more integration test style. In TestPayForBlob, when we submit a PFB with an account that doesn't exist it'll panic unless I have `resp != nil`.

Reply:
> depending on at what point the tx failed the response will include tx hash and error code

Why? If the tx failed, why not include that info in the error. IMO it's really confusing to reason about functions if an error is communicated via the value return argument instead of the error argument. In other words, it is desirable for a function to return an error or a response but never both. This is desirable because callers of the function can propagate errors easily without inspecting the body of the function.